### PR TITLE
Feature/dilution

### PR DESCRIPTION
**Pull request by maddyscientist**
This PR adds dilution functionality to QUDA
* Interface function is `spinorDilute`, which dilutes a fermion field into a vector of fields
* Initial dilution types supported are:
  * spin
  * color
  * spin-color
  * spin-color-even-odd (only applicable to full fields)
* An accompanying regression test is included `dilution_test` which tests that the various dilution patterns are all lossless (e.g., the original vector can be exactly reconstructed), testing both full fields and single parity fields

I also snuck in a compilation fix for nvc++ (which doesn't yet support `__builtin_assume`)

### CODE REVIEW

File: tests/CMakeLists.txt [hunk 1]
```diff
@@ -1049,4 +1055,11 @@ foreach(prec IN LISTS TEST_PRECS)
                      --gtest_output=xml:gauge_arg_test_${prec}.xml)
   endif()
 
+  if (QUDA_NSPIN4)
```
Comment (mathiaswagner) [inline tests/CMakeLists.txt]: Better use 
```suggestion
  if (TARGET dilution_test)
```

File: tests/dilution_test.cpp [hunk 2]
```diff
@@ -0,0 +1,148 @@
+#include <stdlib.h>
```
Comment (mathiaswagner) [inline tests/dilution_test.cpp]: Use C++ version of headers here?

**Comment by weinbe2** [comment]
This is a great, narrow PR, thanks @maddyscientist ! My only ask is if you can extend the test to cover Nspin = 1 as well? It looks like that'd require a new `constructStaggeredTestSpinorParam` function, but that should be relatively straightforward.

File: tests/dilution_test.cpp (hunk 2 above)
Comment (maddyscientist) [reply tests/dilution_test.cpp]: These headers weren't even needed, so I just deleted them 5ca61ad2adaa22da1d40d28ab496bc1dd018ea2b

File: tests/CMakeLists.txt (hunk 1 above)
Comment (maddyscientist) [reply tests/CMakeLists.txt]: Done 5ca61ad2adaa22da1d40d28ab496bc1dd018ea2b

**Comment by maddyscientist** [comment]
Good suggestion @weinbe2 for adding a staggered test: done with 5ca61ad2adaa22da1d40d28ab496bc1dd018ea2b.

Jenkins should now pass as well.

**Review by mathiaswagner — APPROVED** [review]
LGTM. Clean addition.

**Review by weinbe2 — APPROVED** [review]
This looks great to me. Thanks @maddyscientist !

**Review by hummingtree — APPROVED** [review]
This looks nice and clean to me!

**Pull request merged**